# Exercise

In [1]:
import pandas as pd
import numpy as np

df = pd.read_excel("customer_transactions.xlsx", sheet_name="Sheet1")

df.head()

Customer_ID   Name   Age  Gender Education_Level  Number_of_Children  \
0         1001  Alice  43.0    Male       Secondary                 5.0   
1         1002  David  66.0    Male         Diploma                 5.0   
2         1003    Bob  61.0    Male         Diploma                 3.0   
3         1004    Ian  37.0    Male       Secondary                 0.0   
4         1005  Grace  64.0  Female             PhD                 5.0   

      City  Purchase_Amount Purchase_Date Product_Category  \
0  Mombasa           629.59    2023-10-29       Automotive   
1  Mombasa           424.31    2023-04-04         Clothing   
2  Eldoret           220.66    2023-04-10        Furniture   
3    Nyeri            13.95    2023-08-03           Beauty   
4    Thika           666.85    2023-10-07      Electronics   

   Purchase_Frequency  Total_Spent Preferred_Payment_Method  
0                   9      2981.75            Bank Transfer  
1                   2      1397.40              Credit Card  
2                   9      3769.70             Mobile Money  
3                   2      2973.28             Mobile Money  
4                   1       214.64             Mobile Money

# Geting basic information

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1500 entries, 0 to 1499
Data columns (total 13 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   Customer_ID               1500 non-null   int64         
 1   Name                      1440 non-null   object        
 2   Age                       1466 non-null   float64       
 3   Gender                    1500 non-null   object        
 4   Education_Level           1500 non-null   object        
 5   Number_of_Children        1429 non-null   float64       
 6   City                      1456 non-null   object        
 7   Purchase_Amount           1471 non-null   float64       
 8   Purchase_Date             1419 non-null   datetime64[ns]
 9   Product_Category          1500 non-null   object        
 10  Purchase_Frequency        1500 non-null   int64         
 11  Total_Spent               1500 non-null   float64       
 12  Preferred_Payment_Me

In [3]:
print(df.shape)

(1500, 13)


# Statistics summary

In [4]:
print(df.describe())

       Customer_ID          Age  Number_of_Children  Purchase_Amount  \
count  1500.000000  1466.000000          1429.00000      1471.000000   
mean   1750.500000    49.630286             2.96571       463.976540   
std     433.157015    17.829217             2.02459       300.205729   
min    1001.000000    18.000000             0.00000       -49.780000   
25%    1375.750000    34.000000             1.00000       206.645000   
50%    1750.500000    50.500000             3.00000       453.710000   
75%    2125.250000    65.000000             5.00000       724.490000   
max    2500.000000    80.000000             6.00000       999.420000   

       Purchase_Frequency  Total_Spent  
count         1500.000000  1500.000000  
mean             5.309333  2521.933413  
std              2.899526  1424.549468  
min              1.000000   100.420000  
25%              3.000000  1280.882500  
50%              5.000000  2515.110000  
75%              8.000000  3767.222500  
max             10.0000

In [5]:
missing_values = df.isnull().sum()
missing_values

Customer_ID                  0
Name                        60
Age                         34
Gender                       0
Education_Level              0
Number_of_Children          71
City                        44
Purchase_Amount             29
Purchase_Date               81
Product_Category             0
Purchase_Frequency           0
Total_Spent                  0
Preferred_Payment_Method     0
dtype: int64

# Identify data types and any potential inconsistencies

In [6]:
print (df.dtypes)

Customer_ID                          int64
Name                                object
Age                                float64
Gender                              object
Education_Level                     object
Number_of_Children                 float64
City                                object
Purchase_Amount                    float64
Purchase_Date               datetime64[ns]
Product_Category                    object
Purchase_Frequency                   int64
Total_Spent                        float64
Preferred_Payment_Method            object
dtype: object


# Handle Missing Values

In [7]:
#Replace missing values in the Name column with "Unknown".
df['Name'].fillna('Unknown', inplace=True)
df

Customer_ID     Name   Age  Gender      Education_Level  \
0            1001    Alice  43.0    Male            Secondary   
1            1002    David  66.0    Male              Diploma   
2            1003      Bob  61.0    Male              Diploma   
3            1004      Ian  37.0    Male            Secondary   
4            1005    Grace  64.0  Female                  PhD   
...           ...      ...   ...     ...                  ...   
1495         2496      Ian  55.0    Male               Master   
1496         2497      Eve  75.0    Male               Master   
1497         2498      Bob  47.0    Male            Secondary   
1498         2499    David  38.0    Male              Primary   
1499         2500  Charlie  76.0    Male  No formal education   

      Number_of_Children     City  Purchase_Amount Purchase_Date  \
0                    5.0  Mombasa           629.59    2023-10-29   
1                    5.0  Mombasa           424.31    2023-04-04   
2                    3.0  Eldoret           220.66    2023-04-10   
3                    0.0    Nyeri            13.95    2023-08-03   
4                    5.0    Thika           666.85    2023-10-07   
...                  ...      ...              ...           ...   
1495                 1.0    Thika           436.37    2023-02-16   
1496                 NaN  Malindi           650.64    2023-10-10   
1497                 2.0  Garissa           145.98    2023-01-24   
1498                 NaN   Nakuru              NaN    2023-07-19   
1499                 3.0  Eldoret            46.88    2023-03-03   

     Product_Category  Purchase_Frequency  Total_Spent  \
0          Automotive                   9      2981.75   
1            Clothing                   2      1397.40   
2           Furniture                   9      3769.70   
3              Beauty                   2      2973.28   
4         Electronics                   1       214.64   
...               ...                 ...          ...   
1495           Beauty                   3      4652.81   
1496      Electronics                   9       891.92   
1497      Electronics                   5      2048.69   
1498       Automotive                   6      2359.37   
1499             Toys                   4      2569.68   

     Preferred_Payment_Method  
0               Bank Transfer  
1                 Credit Card  
2                Mobile Money  
3                Mobile Money  
4                Mobile Money  
...                       ...  
1495            Bank Transfer  
1496                     Cash  
1497            Bank Transfer  
1498                     Cash  
1499             Mobile Money  

[1500 rows x 13 columns]

In [8]:
#Fill missing values in the Age column with the mean age of customers
df['Age'].fillna(df['Age'].mean(), inplace=True)
df['Age'] = df['Age'].round().astype(int)

In [9]:
#Fill missing values in the Number_of_Children column with 0
df['Number_of_Children'].fillna(0, inplace=True)

In [10]:
#Fill missing values in the City column with "Unknown".
df['City'].fillna('Unknown', inplace=True)

In [11]:
#If Purchase_Date is missing, replace it with "2023-01-01".
df['Purchase_Date'].fillna('2023-01-01', inplace=True)

In [12]:
#If Purchase_Amount is missing, set it to 0
df['Purchase_Amount'].fillna(0, inplace=True)

In [13]:
missing_values = df.isnull().sum()
missing_values

Customer_ID                 0
Name                        0
Age                         0
Gender                      0
Education_Level             0
Number_of_Children          0
City                        0
Purchase_Amount             0
Purchase_Date               0
Product_Category            0
Purchase_Frequency          0
Total_Spent                 0
Preferred_Payment_Method    0
dtype: int64

In [14]:
print (df.dtypes)

Customer_ID                          int64
Name                                object
Age                                  int32
Gender                              object
Education_Level                     object
Number_of_Children                 float64
City                                object
Purchase_Amount                    float64
Purchase_Date               datetime64[ns]
Product_Category                    object
Purchase_Frequency                   int64
Total_Spent                        float64
Preferred_Payment_Method            object
dtype: object


# Identify and Correct Invalid Data

In [15]:
#Some Purchase_Amount values are negative. Set all negative values to 0

#df["Purchase_Amount"] = df["Purchase_Amount"]. clip(lower = 0)
df.loc[df["Purchase_Amount"] < 0, 'Purchase_Amount'] = 0

In [16]:
#neg = df[df["Purchase_Amount"]<0]
#neg

In [17]:
#Verify if there are any customers with age below 18 or above 80. If so, replace them with the median age
#(df[(df['Age'] <18) & (df['Age'] >80)])

df.loc[(df['Age'] < 18) | (df['Age'] > 80), 'Age'] = df['Age'].median()

In [18]:
#Ensure that Purchase_Frequency is between 1 and 10. If outside this range, replace it with 5 (average frequency).
#(df[(df['Purchase_Frequency'] < 1) & (df['Purchase_Frequency'] > 10)])

df.loc[(df['Purchase_Frequency'] < 1) | (df['Purchase_Frequency'] > 10), 'Purchase_Frequency'] = 5

# Variable manipulations

In [19]:
#Create a function that categorizes customers based on Age into:
#"Youth": Age < 30
#"Adult": 30 ≤ Age < 50
#"Senior": Age ≥ 50
#Apply this function to create a new column "Age_Group".

def age_group(age):
        if age < 30:
            return "Youth"
        elif 30 <= age < 50:
            return "Adult"
        else:
            return "Senior"
        
df['Age_Group'] = df['Age'].apply(age_group)

In [20]:
df.head()

Customer_ID   Name  Age  Gender Education_Level  Number_of_Children  \
0         1001  Alice   43    Male       Secondary                 5.0   
1         1002  David   66    Male         Diploma                 5.0   
2         1003    Bob   61    Male         Diploma                 3.0   
3         1004    Ian   37    Male       Secondary                 0.0   
4         1005  Grace   64  Female             PhD                 5.0   

      City  Purchase_Amount Purchase_Date Product_Category  \
0  Mombasa           629.59    2023-10-29       Automotive   
1  Mombasa           424.31    2023-04-04         Clothing   
2  Eldoret           220.66    2023-04-10        Furniture   
3    Nyeri            13.95    2023-08-03           Beauty   
4    Thika           666.85    2023-10-07      Electronics   

   Purchase_Frequency  Total_Spent Preferred_Payment_Method Age_Group  
0                   9      2981.75            Bank Transfer     Adult  
1                   2      1397.40              Credit Card    Senior  
2                   9      3769.70             Mobile Money    Senior  
3                   2      2973.28             Mobile Money     Adult  
4                   1       214.64             Mobile Money    Senior

In [21]:
# •	Create another function that categorizes customers based on their Total_Spent:
#o	"Low Spender": Total Spent < 1000
#o	"Moderate Spender": 1000 ≤ Total Spent < 3000
#o	"High Spender": Total Spent ≥ 3000
#•	Apply this function to create a new column "Spending_Category"


def totalspent_group(spent):
        if spent < 1000:
            return "Low Spender"
        elif 1000 <= spent < 3000:
            return "Moderate Spender"
        else:
            return "High Spender"
        
df['Spending_Category'] = df['Total_Spent'].apply(totalspent_group)

In [22]:
df.head()

Customer_ID   Name  Age  Gender Education_Level  Number_of_Children  \
0         1001  Alice   43    Male       Secondary                 5.0   
1         1002  David   66    Male         Diploma                 5.0   
2         1003    Bob   61    Male         Diploma                 3.0   
3         1004    Ian   37    Male       Secondary                 0.0   
4         1005  Grace   64  Female             PhD                 5.0   

      City  Purchase_Amount Purchase_Date Product_Category  \
0  Mombasa           629.59    2023-10-29       Automotive   
1  Mombasa           424.31    2023-04-04         Clothing   
2  Eldoret           220.66    2023-04-10        Furniture   
3    Nyeri            13.95    2023-08-03           Beauty   
4    Thika           666.85    2023-10-07      Electronics   

   Purchase_Frequency  Total_Spent Preferred_Payment_Method Age_Group  \
0                   9      2981.75            Bank Transfer     Adult   
1                   2      1397.40              Credit Card    Senior   
2                   9      3769.70             Mobile Money    Senior   
3                   2      2973.28             Mobile Money     Adult   
4                   1       214.64             Mobile Money    Senior   

  Spending_Category  
0  Moderate Spender  
1  Moderate Spender  
2      High Spender  
3  Moderate Spender  
4       Low Spender

# Use Loops and Conditional Statements

In [23]:
#Count and print the number of customers in each "Age_Group".
#Count and print the number of customers in each "Spending_Category"

categorical_columns = ["Age_Group", "Spending_Category"]
for col in categorical_columns:
    print(f"\nUnique values in {col}:")
    print(df[col].value_counts())


Unique values in Age_Group:
Senior    781
Adult     461
Youth     258
Name: Age_Group, dtype: int64

Unique values in Spending_Category:
Moderate Spender    615
High Spender        605
Low Spender         280
Name: Spending_Category, dtype: int64


In [24]:
# Write another loop to count the number of customers in each "Spending_Category", using an if-elif-else statement to check the category.
spending_counts = {"Low Spender": 0, "Moderate Spender": 0, "High Spender": 0}
for spending_category in df["Spending_Category"].astype(str):  
    if spending_category == "Low Spender":
        spending_counts["Low Spender"] += 1
    elif spending_category == "Moderate Spender":
        spending_counts["Moderate Spender"] += 1
    elif spending_category == "High Spender":
        spending_counts["High Spender"] += 1
    else:
        print(f"Unexpected category found: {spending_category}")
print("\nCustomer count by Spending Category:")
for category, count in spending_counts.items():
    print(f"{category}: {count}")


Customer count by Spending Category:
Low Spender: 280
Moderate Spender: 615
High Spender: 605


In [25]:
# Find and print all customers who have made more than 5 purchases and have spent more than 3000 total (a high-value customer segment).
print("\nHigh-Value Customers:")
for index, row in df.iterrows():
    if row["Purchase_Frequency"] > 5 and row["Total_Spent"] > 3000:
        print(f"Customer ID: {row['Customer_ID']}, Name: {row['Name']}, Total Spent: {row['Total_Spent']}")


High-Value Customers:
Customer ID: 1003, Name: Bob, Total Spent: 3769.7
Customer ID: 1012, Name: Charlie, Total Spent: 4969.21
Customer ID: 1014, Name: David, Total Spent: 4247.79
Customer ID: 1018, Name: Alice, Total Spent: 3602.86
Customer ID: 1020, Name: Grace, Total Spent: 4803.17
Customer ID: 1021, Name: Eve, Total Spent: 4237.77
Customer ID: 1026, Name: Eve, Total Spent: 3905.48
Customer ID: 1027, Name: David, Total Spent: 4728.03
Customer ID: 1028, Name: Grace, Total Spent: 4422.37
Customer ID: 1035, Name: Grace, Total Spent: 4503.89
Customer ID: 1043, Name: Frank, Total Spent: 4066.98
Customer ID: 1045, Name: Hannah, Total Spent: 4046.27
Customer ID: 1049, Name: Jane, Total Spent: 4435.21
Customer ID: 1053, Name: Bob, Total Spent: 3617.78
Customer ID: 1054, Name: Charlie, Total Spent: 4497.56
Customer ID: 1060, Name: Bob, Total Spent: 3506.2
Customer ID: 1063, Name: Grace, Total Spent: 3901.25
Customer ID: 1075, Name: Frank, Total Spent: 3261.04
Customer ID: 1077, Name: David,

In [26]:
# Count of all customers who have made more than 5 purchases and have spent more than 3000 total (a high-value customer segment).
high_value_count = df[(df["Purchase_Frequency"] > 5) & (df["Total_Spent"] > 3000)].shape[0]
print(f"\nTotal High-Value Customers: {high_value_count}")


Total High-Value Customers: 282


In [27]:
# Identify customers who have missing names and print their Customer_IDs.
print("\nCustomers with missing names:")
for index, row in df.iterrows():
    if pd.isna(row["Name"]) or str(row["Name"]).strip() == "Unknown":
        print(f"Customer ID: {row['Customer_ID']}")


Customers with missing names:
Customer ID: 1006
Customer ID: 1025
Customer ID: 1092
Customer ID: 1137
Customer ID: 1145
Customer ID: 1155
Customer ID: 1168
Customer ID: 1202
Customer ID: 1233
Customer ID: 1313
Customer ID: 1404
Customer ID: 1422
Customer ID: 1462
Customer ID: 1470
Customer ID: 1509
Customer ID: 1521
Customer ID: 1538
Customer ID: 1560
Customer ID: 1579
Customer ID: 1635
Customer ID: 1637
Customer ID: 1640
Customer ID: 1651
Customer ID: 1659
Customer ID: 1665
Customer ID: 1676
Customer ID: 1716
Customer ID: 1745
Customer ID: 1781
Customer ID: 1818
Customer ID: 1849
Customer ID: 1852
Customer ID: 1905
Customer ID: 1907
Customer ID: 1940
Customer ID: 1960
Customer ID: 1981
Customer ID: 2021
Customer ID: 2045
Customer ID: 2054
Customer ID: 2058
Customer ID: 2097
Customer ID: 2114
Customer ID: 2150
Customer ID: 2190
Customer ID: 2194
Customer ID: 2196
Customer ID: 2199
Customer ID: 2226
Customer ID: 2241
Customer ID: 2253
Customer ID: 2293
Customer ID: 2311
Customer ID: 23

# Grouping and Aggregation

In [28]:
#Find the total Purchase_Amount spent per "Age_Group", and display the results
purchase_total = df.groupby("Age_Group")["Purchase_Amount"].sum() 
purchase_total

Age_Group
Adult     206740.27
Senior    359684.59
Youth     118043.43
Name: Purchase_Amount, dtype: float64

In [29]:
#Find the average Purchase_Amount per "Product_Category".
product_average = df.groupby("Product_Category")["Purchase_Amount"].mean() 
product_average

Product_Category
Automotive     460.037471
Beauty         423.380000
Books          443.296627
Clothing       480.814492
Electronics    461.627673
Furniture      453.539277
Groceries      451.337938
Sports         470.256795
Toys           460.340000
Name: Purchase_Amount, dtype: float64

In [30]:
# Find the most common "Preferred_Payment_Method" among High Spenders.
preferred_method = df[df["Spending_Category"] == "High Spender"]["Preferred_Payment_Method"].mode()[0]
preferred_method

'Credit Card'

# Additional exercise

In [31]:
#Find the top 5 cities with the highest total spending

highest_city_spending = df.groupby("City")["Total_Spent"].sum().nlargest(5)
highest_city_spending

City
Nakuru     400763.73
Mombasa    384398.25
Malindi    375804.00
Nyeri      373292.12
Garissa    369293.19
Name: Total_Spent, dtype: float64

In [32]:
#Find the average number of children for each "Education_Level".
average_educational_level = df.groupby("Education_Level")["Number_of_Children"].mean() 
average_educational_level

Education_Level
Bachelor               2.768844
Diploma                2.861345
Master                 2.826087
No formal education    2.911628
PhD                    2.633028
Primary                2.871134
Secondary              2.907767
Name: Number_of_Children, dtype: float64

In [33]:
#Identify the "Product_Category" with the highest total revenue
highest_product_category = df.groupby("Product_Category")["Total_Spent"].sum().nlargest(1)
highest_product_category

Product_Category
Clothing    483596.16
Name: Total_Spent, dtype: float64

In [34]:
# Save the cleaned dataset in excel
df.to_excel(r"C:\Users\Gitamo Adrian\Desktop\Groups 1 and 3\cleaned_data.xlsx")

In [35]:
df.columns

Index(['Customer_ID', 'Name', 'Age', 'Gender', 'Education_Level',
       'Number_of_Children', 'City', 'Purchase_Amount', 'Purchase_Date',
       'Product_Category', 'Purchase_Frequency', 'Total_Spent',
       'Preferred_Payment_Method', 'Age_Group', 'Spending_Category'],
      dtype='object')

# Final Step
# Write a short summary of key insights.

## Key insights:

- Spending Category:

    - Low Spender: 280
    - Moderate Spender: 615
    - High Spender: 605
    - Age Group:
    - Adult: 461
    - Senior: 781
    - Youth: 258

- There were no customers with age below 18 or above 80.
- The Purchase_Frequency was between 1 and 10 and none was found outside this range.

- Key Customer Segments:

    - High-Value Customers (more than 5 purchases and spent over 3000): 282
    - Customers with Missing or 'Unknown' Names: 60

- Total Purchase Amount by Age Group:

    - Adult: 206,740.27
    - Senior: 359,684.59
    - Youth: 118,043.43
    
- Average Purchase Amount by Product Category:

    - Automotive     460.037471
    - Beauty         423.380000
    - Books          443.296627
    - Clothing       480.814492 (highest total revenue)
    - Electronics    461.627673
    - Furniture      453.539277
    - Groceries      451.337938
    - Sports         470.256795
    - Toys           460.340000

  
- Preferred Payment Method among High Spenders: Credit Card

- Top 5 Cities by Total Spending:

    - Nakuru: 400,763.73
    - Mombasa: 384,398.25
    - Malindi: 375,804.00
    - Nyeri: 373,292.12
    - Garissa: 369,293.19

  
- Average Number of Children by Education Level:

    - Bachelor               2.768844
    - Diploma                2.861345
    - Master                 2.826087
    - No formal education    2.911628
    - PhD                    2.633028
    - Primary                2.871134
    - Secondary              2.907767